### DISCUSSION TITLE
Affordable programmer-debugger probe by Seeed

### DISCUSSION THREAD

**Discussion by rei-vilo** [Seeed-Studio/OSHW-XIAO-Series#20] [In Development]
A programmer-debugger probe is a very useful and "must have" tool for embedded system development.

Today, there is no affordable offer.

Segger used to list the [J-Link Edu mini](https://www.segger.com/products/debug-probes/j-link/models/j-link-edu-mini/) probe at an affordable price; it is no longer the case. Price went from USD20 to [USD70](https://octopart.com/search?q=segger+j-link+edu+mini&currency=USD).

The Xiao expansion board provides the pogo pins to make the connections, alas only for some boards, SAMD, nRF52840 and RP2040. The ESP32 requires full JTAG, except the latest C-series which include the debugger.

I am not a bing fan of the fragile pogo pins featured by the [Xiao Expansion Board](https://www.seeedstudio.com/Seeeduino-XIAO-Expansion-board-p-4746.html). Once the Xiao is mounted in the final design, the SWD pads are no longer accessible unless they are rerouted. The 2x5 0.05" connector is very compact, and some variants even include power and UART RX/TX signals.

I would replace the USB connector by a SWD connector on the Xiao boards. The USB connector takes too much space and is of limited use. 

The [Xiao SAMD turned debugger](https://embeddedcomputing.weebly.com/the-5-programmer-debugger.html) provides DAPLink but does not support all the Xiao boards. 

Seeed could take inspiration of the [Raspberry Pi Debug Probe](https://www.raspberrypi.com/documentation/microcontrollers/debug-probe.html) brings the best-in-class value, with its fool-proof connector, software tools suite and UART RX/TX signals

References:
* [Roadmap of Seeed XIAO: You Decide What We Build Next!](https://github.com/Seeed-Studio/OSHW-XIAO-Series/discussions/1#discussioncomment-10095392)
* [Why should we consider redesign the SWD interface on the XIAO](https://forum.seeedstudio.com/t/why-should-we-consider-redesign-the-swd-interface-on-the-xiao/280182/5?u=reivilo)
* [The $5 programmer-debugger](https://embeddedcomputing.weebly.com/the-5-programmer-debugger.html)

**Comment by limengdu** [comment]
OpenOCD, BlackMagic Probe, DAPLink, PyOCD are a few open source solutions, which one do you think is better for us to use?

**Reply by rei-vilo** [reply]
I would go for OpenOCD, as I am using it against the boards I work with. Now, I have no experience with BlackMagic.

**Comment by limengdu** [comment]
![正面@1x](https://github.com/user-attachments/assets/78d3b9a2-895d-4452-a687-15792a6c4aee)
![背面@1x](https://github.com/user-attachments/assets/f0357480-1058-4301-9e23-2645490bd0a9)
preliminary design

**Reply by rei-vilo** [reply]
@limengdu Thank you for sharing the design. As always, some questions and some comments.

* What are the features of the board? Does it include an MCU for the toggle wheel, the display and the debugger? 
* How to connect the board to a PC?
* Where is placed or connected the device under test? Does the connection rely on pgo pins?
* Is the board compatible with the new 14 + 9 form-factor of the Xiao Plus?
* I would recommend placing all the connectors, including standard IOs, extended IOs and debug pins, on the same side.  
* Grove UART is fine, but what about a Grove I²C connector as well?
* I am not a big fan of RGB LEDs. I would trade them for another feature to match the targeted budget.

**Reply by CWGwaltney** [reply]
i think it is important to build the XIAO ecosystem... and to implement as many features on a board as the GPIO will permit

**Reply by limengdu** [reply]
Let me answer your questions:
1. This Debug Tool integrates SWD/JTAG debugger function, UART to TTL function and power consumption statistics, focusing on the development and debugging scenarios of all XIAO development boards.
2. Development board built-in USB-C interface, can be directly connected to the computer via cable.
3.XIAO and Debug Tool are directly connected through the motherboard, no other wiring troubles.
4.Support. Correction, it's a 14+8 pin design, and will lead the extra 8 GPIOs out through the row mother.
5. There may not be enough space on one side, currently XIAO is planning a consistent design language, we will use some development board standard size to design XIAO Gadget, but we will try to place the interfaces that may be used on the same side.
6. Please note that the UART is designed to match the TTL function of the Debug Tool, it is not an expansion board, and the IIC does not play any role for this board.
7. This is a single-colour LED, the cost will not be particularly affected. My vision is a very beautiful desktop ornament, lighting effects we will design extremely cool, with the choice of baud rate, I believe you will like such a delicate gadget.

**Comment by limengdu** [comment]
![企业微信截图_1739772671361](https://github.com/user-attachments/assets/eb690a90-0eb7-4af8-8b75-fbabb56d21e7)
The new progress, perhaps, ends up looking like this.

**Reply by rei-vilo** [reply]
Any new development on this project?

Thanks!

**Reply by limengdu** [reply]
The software for this product is more challenging than expected, and we are still working hard to overcome certain technical difficulties

**Reply by rei-vilo** [reply]
Thank you for the update and good luck with the development!

**Comment by limengdu** [comment]
![wecom-temp-2086798-d11ee71da8fb19f9d40cd69788924345](https://github.com/user-attachments/assets/53434292-88a5-4b1b-8d3d-fd5703cc7b1c)
Software development in progress

**Reply by rei-vilo** [reply]
Very nice!

**Comment by limengdu** [comment]
Updated with a video!
https://x.com/seeedstudio/status/1934489411682467899

**Comment by limengdu** [comment]
<img width="1236" alt="image" src="https://github.com/user-attachments/assets/e0587b88-3d71-43fe-a055-07b7877d630e" />
After one more revision, it should be ready to officially meet everyone!

**Reply by rei-vilo** [reply]
Very nice!

**Comment by limengdu** [comment]
This is really the final step before launch! We'd like to collect product name suggestions from everyone. We want it to be intuitive enough so you'll immediately recognize it as a multi-functional debugging tool. We want it to be easily searchable, making it more convenient for you to find. We've brainstormed many options: Debugger for XIAO, Multi-Tool for XIAO, XIAO Multi-Tool... but we're finding it hard to decide. Do you have any good suggestions or ideas? What keywords would you search for when looking for similar tools? Thanks in advance!

**Reply by rei-vilo** [reply]
As I haven’t tested the Xiao probe yet, I don’t have many ideas for a name.

**Comment by meilily** [comment]
Hey community, a big piece of news to share that XIAO Debug Mate is now officially live. Thank you so much for all your discussion and contributions to the development of this product. 

 [Product Page](https://www.seeedstudio.com/Seeed-Studio-XIAO-Debug-Mate-p-6588.html)
 [Full story of how we co-create this product with the community](https://www.seeedstudio.com/blog/2025/11/27/xiao-debug-mate-a-plug-and-play-dev-multi-tool-for-19-9/)

![XIAO Debug Mate](https://github.com/user-attachments/assets/38660d02-a2c1-485e-a920-ee8df2dddc6f)
